File: Examples/test-suite/csharp/allprotected_runme.cs
using System;
using allprotectedNamespace;

public class runme
{
  static void Main() 
  {
    runme r = new runme();
    r.run();
  }

  void run()
  {
    MyProtectedBase mpb = new MyProtectedBase("MyProtectedBase");
    mpb.accessProtected();
    try {
      // C++ destructor is protected
      mpb.Dispose();
      throw new Exception("failed to catch MethodAccessException");
    } catch (MethodAccessException) {
      // Exception message: C++ destructor does not have public access
    }
    ProtectedDerived pd = new ProtectedDerived("ProtectedDerived");
    // Destroying via the ProtectedDerived's destructor should work
    pd.Dispose();

    ProtectedBase pb = new ProtectedDerived("ProtectedDerived");
    // ProtectedDerived's destructor should be called via the Dispose(disposing) virtual call
    pb.Dispose();
  }
}

class MyProtectedBase : ProtectedBase
{
  public MyProtectedBase(string name) : base(name) {
  }
  public void accessProtected() {
    string s = virtualMethod();
    if (s != "ProtectedBase")
      throw new Exception("Failed");

    Klass k = instanceMethod(new Klass("xyz"));
    if (k.getName() != "xyz") 
      throw new Exception("Failed");

    k = instanceOverloaded(new Klass("xyz"));
    if (k.getName() != "xyz") 
      throw new Exception("Failed");

    k = instanceOverloaded(new Klass("xyz"), "abc");
    if (k.getName() != "abc") 
      throw new Exception("Failed");

    k = ProtectedBase.staticMethod(new Klass("abc"));
    if (k.getName() != "abc") 
      throw new Exception("Failed");

    k = ProtectedBase.staticOverloaded(new Klass("xyz"));
    if (k.getName() != "xyz") 
      throw new Exception("Failed");

    k = ProtectedBase.staticOverloaded(new Klass("xyz"), "abc");
    if (k.getName() != "abc") 
      throw new Exception("Failed");

    instanceMemberVariable = 30;
    int i = instanceMemberVariable;
    if (i != 30)
      throw new Exception("Failed");

    staticMemberVariable = 40;
    i = staticMemberVariable;
    if (i != 40)
      throw new Exception("Failed");

    i = staticConstMemberVariable;
    if (i != 20)
      throw new Exception("Failed");

    anEnum = ProtectedBase.AnEnum.EnumVal1;
    ProtectedBase.AnEnum ae = anEnum;
    if (ae != ProtectedBase.AnEnum.EnumVal1)
      throw new Exception("Failed");
  }
}

File: Examples/test-suite/lua/allprotected_runme.lua
local v=require("allprotected")

local klass = v.Klass("allprotected_klass")
assert(klass:getName() == "allprotected_klass", "klass.getName()")

local publicBase = v.PublicBase("allprotected_PubBase")
assert(publicBase:virtualMethod() == "PublicBase", "publicBase.virtualMethod()")

local class = publicBase:instanceMethod(klass)
assert(class:getName() == "allprotected_klass", "publicBase.instanceMethod(klass).getName()")

class = publicBase:instanceOverloaded(klass)
assert(class:getName() == "allprotected_klass", "publicBase.instanceOverloaded(klass).getName()")

class = v.PublicBase.staticMethod(klass)
assert(class:getName() == "allprotected_klass", "PublicBase.staticMethod(klass).getName()")

class = v.PublicBase.staticOverloaded(klass)
assert(class:getName() == "allprotected_klass", "PublicBase.staticOverloaded(klass).getName()")

class = v.PublicBase.staticOverloaded(klass, "allprotected_klass3")
assert(class:getName() == "allprotected_klass3", "PublicBase.staticOverloaded(klass, \"allprotected_klass3\").getName()")

assert(v.PublicBase.EnumVal1 == 0, "PublicBase.EnumVal1")
assert(v.PublicBase.EnumVal2 == 1, "PublicBase.EnumVal2")

v.PublicBase.instanceMemberVariable = 12
assert(v.PublicBase.instanceMemberVariable == 12, "PublicBase.instanceMemberVariable")

assert(v.PublicBase.staticConstMemberVariable == 20, "PublicBase.staticConstMemberVariable")
assert(v.PublicBase.staticMemberVariable == 10, "PublicBase.staticMemberVariable")

publicBase.stringMember = "dummy"
assert(publicBase.stringMember == "dummy", "publicBase.stringMember")

assert(publicBase.anEnum == v.PublicBase.EnumVal1, "publicBase.anEnum == EnumVal1");
publicBase.anEnum = v.PublicBase.EnumVal2
assert(publicBase.anEnum == v.PublicBase.EnumVal2, "publicBase.anEnum == EnumVal2");

local procBase = v.ProtectedBase("allprotected_ProcBase")
assert(procBase ~= nil)
assert(procBase:virtualMethod() == "ProtectedBase", "procBase.virtualMethod()")
assert(procBase.anEnum == v.ProtectedBase.EnumVal1, "procBase.anEnum")
